### PR TITLE
Add OW_UNION_DISABLE_CHECK and OW_UNION_CONNECTED_ONLY_CHECK

### PR DESCRIPTION
**Pull request by pkmnsnfrn**
# Description
Player enters Pokemon Center without lag

* Adds `OW_UNION_DISABLE_CHECK`. When TRUE, the nurse does not inform the player if there is a trainer waiting in the Union Room or at Trainer Hill. Entering a Pok├⌐mon Center will be faster as a result.
* Adds `OW_UNION_CONNECTED_ONLY_CHECK`. When TRUE, the Union Room check only occurs unless the Wireless Adapter is connected.

#### `OW_UNION_DISABLE_CHECK`
When this is `TRUE`:
* `InitUnionRoom` does not run when entering a Pokemon Center.
* `EventScript_PkmnCenterNurse_CheckTrainerHillAndUnionRoom` will never run.

#### `OW_UNION_CONNECTED_ONLY_CHECK`
When this is `TRUE`:
* `IsWirelessAdapterConnected` is run. 
* If the adapter is NOT connected, `InitUnionRoom` does not run.
* If the adapter is NOT connected, `FLAG_NURSE_UNION_ROOM_REMINDER` is set.
	* When `FLAG_NURSE_UNION_ROOM_REMINDER` is set, `BufferUnionRoomPlayerName` will always return `FALSE`.
	* When `FLAG_NURSE_UNION_ROOM_REMINDER` is set, `EventScript_PkmnCenterNurse_CheckTrainerHillAndUnionRoom` will goto `EventScript_PkmnCenterNurse_ReturnPkmn`.

# Testing
## Clean Branch
You can recreate this branch by applying a patch or pulling the repo. From a clean version of expansion's upcoming, you can either:

### Patch
`wget https://files.catbox.moe/e9sc6e.patch -O union.patch ; git apply union.patch ; rm union.patch`

### Repo
`git remote add psf-expansion https://github.com/PokemonSanFran/pokeemerald-expansion/ ; git pull psf-expansion noAdapterUnionCheck`

## Manual Tests
After replicating the branch, to recreate my testing environment, you can either directly download the debug script, or manually create the changes.

#### `OW_UNION_DISABLE_CHECK` == `FALSE` && `OW_UNION_CONNECTED_ONLY_CHECK ` == `FALSE`
`wget https://files.catbox.moe/vg4kar.h -O include/config/overworld.h`
#### `OW_UNION_DISABLE_CHECK` == `FALSE` && `OW_UNION_CONNECTED_ONLY_CHECK ` == `TRUE`
`wget https://files.catbox.moe/h4enwo.h -O include/config/overworld.h`
#### `OW_UNION_DISABLE_CHECK` == `TRUE` && `OW_UNION_CONNECTED_ONLY_CHECK ` == `FALSE`
`wget https://files.catbox.moe/s25anv.h -O include/config/overworld.h`
#### `OW_UNION_DISABLE_CHECK` == `TRUE` && `OW_UNION_CONNECTED_ONLY_CHECK ` == `TRUE`
`wget https://files.catbox.moe/bnzfaw.h -O include/config/overworld.h`

### Manual Testing
* Modify `include/config/overworld.h` to the desired values
* Compile, start a new save
* Run Cheat Start
* Leave the truck
* Enter the Center

## Verified Scenarios
All videos start from a Player that has used Cheat Start and saved outside of the Pokemon Center.
* Walk into Pokemon Center
* Heal Pokemon with Nurse Joy
* Walk outside of Pokemon Center
* Walk into Pokemon Center

#### `OW_UNION_DISABLE_CHECK` == `FALSE` && `OW_UNION_CONNECTED_ONLY_CHECK ` == `FALSE`
https://github.com/user-attachments/assets/cfa9d622-11e6-421e-8252-39a7bf8f6764
#### `OW_UNION_DISABLE_CHECK` == `FALSE` && `OW_UNION_CONNECTED_ONLY_CHECK ` == `TRUE`
https://github.com/user-attachments/assets/69a92ba3-509f-407d-b2b2-65ca359674e8
#### `OW_UNION_DISABLE_CHECK` == `TRUE` && `OW_UNION_CONNECTED_ONLY_CHECK ` == `FALSE`
https://github.com/user-attachments/assets/7a3777f5-786c-46be-b1d1-541115f5d015
#### `OW_UNION_DISABLE_CHECK` == `TRUE` && `OW_UNION_CONNECTED_ONLY_CHECK ` == `TRUE`
https://github.com/user-attachments/assets/d66700e9-4230-4ac3-bdc1-b0cf81fffbab

# **People who collaborated with me in this PR**
This was originally [written by](<https://github.com/pret/pokeemerald/wiki/Disabling-Union-Room-check-when-entering-Pok%C3%A9mon-Centers>) the combined efforts of @Deokishisu, @LOuroboros, @Temporalin, @GriffinRichards, @Jaizu. Give them all the credit.

# Features this PR does NOT handle:

# Discord Contact Info
I am `pkmnsnfrn` on Discord.

### CODE REVIEW

File: data/scripts/cable_club.inc [hunk 1]
```diff
@@ -1257,7 +1257,14 @@ EventScript_CloseMossdeepGameCornerBarrier::
 	return
 
 CableClub_OnResume:
+.if OW_UNION_CONNECTED_ONLY_CHECK == TRUE
```
Comment (Bassoonian) [inline data/scripts/cable_club.inc:1260]: If `OW_UNION_DISABLE_CHECK` is true, this entire check is pointless because it goes to `CableClub_OnResume_End` either way. I would personally only run this portion of the code if `OW_UNION_DISABLE_CHECK` is false (you could embed it inside a bigger `.if OW_UNION_DISABLE_CHECK == FALSE`

File: include/config/overworld.h [hunk 2]
```diff
@@ -95,6 +95,8 @@
 #define OW_POPUP_BW_ALPHA_BLEND    FALSE                    // Enables alpha blending/transparency for the pop-ups. Mainly intended to be used with the black color option.
 
 // Pokémon Center
-#define OW_IGNORE_EGGS_ON_HEAL     GEN_LATEST               // In Gen 4+, the nurse in the Pokémon Center does not heal Eggs on healing machine.
+#define OW_IGNORE_EGGS_ON_HEAL                GEN_LATEST    // In Gen 4+, the nurse in the Pokémon Center does not heal Eggs on healing machine.
+#define OW_UNION_DISABLE_CHECK                FALSE         // When TRUE, the nurse does not inform the player if there is a trainer waiting in the Union Room or at Trainer Hill. Entering a Pokémon Center will be faster as a result.
+#define OW_UNION_CONNECTED_ONLY_CHECK         FALSE         // When TRUE, the Union Room check only occurs unless the Wireless Adapter is connected.
```
Comment (Bassoonian) [inline include/config/overworld.h:100]: Might be worth explicitly mentioning this setting is useless if `OW_UNION_DISABLE_CHECK` is true (maybe even a static assert? might be overkill, though)

File: src/link.c [hunk 3]
```diff
@@ -247,6 +247,7 @@ bool8 IsWirelessAdapterConnected(void)
     SetWirelessCommType0_Internal();
     CloseLink();
     RestoreSerialTimer3IntrHandlers();
+    FlagSet(FLAG_NURSE_UNION_ROOM_REMINDER);
```
Comment (Bassoonian) [inline src/link.c:250]: Is this change necessary?

File: src/union_room.c [hunk 4]
```diff
@@ -3377,6 +3377,9 @@ static void Task_InitUnionRoom(u8 taskId)
 
 bool16 BufferUnionRoomPlayerName(void)
 {
+    if (FlagGet(FLAG_NURSE_UNION_ROOM_REMINDER))
+        return FALSE;
```
Comment (Bassoonian) [inline src/union_room.c:3381]: Is this change necessary?

File: src/link.c (hunk 3 above)
Comment (pkmnsnfrn) [reply src/link.c:250]: Answering https://github.com/rh-hideout/pokeemerald-expansion/pull/5050#discussion_r1694220956 as well

When this is run for `OW_UNION_CONNECTED_ONLY_CHECK`, that flag is what prevents the nurse from trying to tell you about it. However, because Game Freak hates us, the game will still try to copy the name of a player from the Union Room. Because the Union Room was never initialized, it will try to do a StringCopy from an empty place, which will crash the game. Returning FALSE prevents that crash.

File: src/union_room.c (hunk 4 above)
Comment (pkmnsnfrn) [reply src/union_room.c:3381]: See https://github.com/rh-hideout/pokeemerald-expansion/pull/5050#discussion_r1694276627

**Comment by Deokishisu** [comment]
While I haven't updated my tutorial on the pokeemerald wiki yet, I now hide the `InitUnionRoom()` call in the nurse's healing field effect animation when the Wireless Adapter is connected and it only is run once per map load with a temp flag. This prevents the map loading lag in both situations and moves the Union Room check to right before it is needed, so the check (and lag) doesn't happen if the player is just stopping by to use the PC or something. There is a small hitch as the nurse puts the Pokeballs on the machine when the Union Room is being initialized, but it is mostly hidden in the field effect animation. I also move the related calls and flag checks to C to make the lag shorter.

[This CrystalDust commit](https://github.com/Deokishisu/CrystalDust/commit/7388c921a41962e042dda63f13d76372df22bc7f#diff-628660f8de2f9b363f82426bdf78cfcd6916e0cfca7717879bccd134c68da78d) shows the changes needed to accomplish this. If anyone can test if the animation lag happens with an actual Wireless Adapter connected on hardware and report back to me (I do not have one and mGBA doesn't emulate it), that would be great.

File: data/scripts/cable_club.inc (hunk 1 above)
Comment (pkmnsnfrn) [reply data/scripts/cable_club.inc:1261]: https://github.com/rh-hideout/pokeemerald-expansion/pull/5050/commits/86593c504a1e65963e8f8e705460460726641a15

File: include/config/overworld.h (hunk 2 above)
Comment (pkmnsnfrn) [reply include/config/overworld.h]: https://github.com/rh-hideout/pokeemerald-expansion/pull/5050/commits/79e4633f789cc877374ce27a3c2ecc50cc585d5f

**Comment by pkmnsnfrn** [comment]
> While I haven't updated my tutorial on the pokeemerald wiki yet, I now hide the `InitUnionRoom()` call in the nurse's healing field effect animation when the Wireless Adapter is connected and it only is run once per map load with a temp flag. This prevents the map loading lag in both situations and moves the Union Room check to right before it is needed, so the check (and lag) doesn't happen if the player is just stopping by to use the PC or something when the adapter is connected. There is a small hitch as the nurse puts the Pokeballs on the machine when the Union Room is being initialized, but it is mostly hidden in the field effect animation. I also move the related calls and flag checks to C to make the lag shorter.
> 
> [This CrystalDust commit](https://github.com/Deokishisu/CrystalDust/commit/7388c921a41962e042dda63f13d76372df22bc7f#diff-628660f8de2f9b363f82426bdf78cfcd6916e0cfca7717879bccd134c68da78d) shows the changes needed to accomplish this. If anyone can test if the animation lag happens with an actual Wireless Adapter connected on hardware and report back to me (I do not have one and mGBA doesn't emulate it), that would be great.

This is certainly more elegant than my current PR, but it costs a flag. I imagine a flag less version with the same end result would have an easier time getting merged from a Senate perspective

**Comment by Deokishisu** [comment]
> > While I haven't updated my tutorial on the pokeemerald wiki yet, I now hide the `InitUnionRoom()` call in the nurse's healing field effect animation when the Wireless Adapter is connected and it only is run once per map load with a temp flag. This prevents the map loading lag in both situations and moves the Union Room check to right before it is needed, so the check (and lag) doesn't happen if the player is just stopping by to use the PC or something when the adapter is connected. There is a small hitch as the nurse puts the Pokeballs on the machine when the Union Room is being initialized, but it is mostly hidden in the field effect animation. I also move the related calls and flag checks to C to make the lag shorter.
> > [This CrystalDust commit](https://github.com/Deokishisu/CrystalDust/commit/7388c921a41962e042dda63f13d76372df22bc7f#diff-628660f8de2f9b363f82426bdf78cfcd6916e0cfca7717879bccd134c68da78d) shows the changes needed to accomplish this. If anyone can test if the animation lag happens with an actual Wireless Adapter connected on hardware and report back to me (I do not have one and mGBA doesn't emulate it), that would be great.
> 
> This is certainly more elegant than my current PR, but it costs a flag. I imagine a flag less version with the same end result would have an easier time getting merged from a Senate perspective

The flag can be reused for other `ON_RESUME` shenanigans. You could also, instead of directly using an `FLAG_SYS_ON_RESUME` flag like I did, change it to a temp flag and maintain a `MAP_SCRIPT_ON_RESUME` in every Pokemon Center like vanilla and have it clear the temp flag. This maintains the original functionality at the cost of still needing a `MAP_SCRIPT_ON_RESUME` in every Pokemon Center.

**Comment by pkmnsnfrn** [comment]
Did some thinking here and I think I'll stick with the original implementation. `Task_InitUnionRoom` is ONLY called when the player enters the Pokemon Center, and I suspect it will not work without it.

Because the check happens on the bow, the only way for the player to actually use the Union Room would be to heal their Pokemon with the nurse, which is clearly a regression.

In the existing implementation, a user, with very little effort, would allow players to still use the Union Room without healing AND avoid the lag associated when the Adapater is NOT plugged in.

File: data/scripts/cable_club.inc [hunk 5]
```diff
@@ -1257,7 +1257,14 @@ EventScript_CloseMossdeepGameCornerBarrier::
 	return
 
 CableClub_OnResume:
+.if OW_UNION_DISABLE_CHECK == FALSE
+.if OW_UNION_CONNECTED_ONLY_CHECK == TRUE
+	specialvar VAR_0x8001, IsWirelessAdapterConnected
+	goto_if_eq VAR_0x8001, FALSE, CableClub_OnResume_End
+.endif
 	special InitUnionRoom
+.endif
```
Comment (Bassoonian) [inline data/scripts/cable_club.inc:1266]:
```suggestion
.endif @ OW_UNION_CONNECTED_ONLY_CHECK == TRUE
	special InitUnionRoom
.endif @ OW_UNION_DISABLE_CHECK == FALSE
```
Just for clarity because of stacking .ifs

**Review by Bassoonian — DISMISSED** [review]

**Comment by Deokishisu** [comment]
> Did some thinking here and I think I'll stick with the original implementation. `Task_InitUnionRoom` is ONLY called when the player enters the Pokemon Center, and I suspect it will not work without it.
> 
> Because the check happens on the bow, the only way for the player to actually use the Union Room would be to heal their Pokemon with the nurse, which is clearly a regression.
> 
> In the existing implementation, a user, with very little effort, would allow players to still use the Union Room without healing AND avoid the lag associated when the Adapater is NOT plugged in.

I don't think this reasoning makes sense. If the player saves in the 2nd floor of the Pokemon Center and resets, they can still enter the Union Room without going downstairs to run `Task_InitUnionRoom`. While I haven't tested the Union Room at all this or last decade (I no longer have the hardware and mGBA doesn't emulate the Wireless Adapter), I know for a fact that I have saved and reset in the 2F Pokemon Center and entered the Union Room without first going downstairs to init it as a kid.

I think that task is only for the nurse's dialogue. If anyone is able to actually test the Union Room to see if `Task_InitUnionRoom` is necessary to enter by saving and resetting in the 2F and trying to enter the Union Room without going downstairs, that would confirm it.

**Comment by pkmnsnfrn** [comment]
@Bassoonian Deokishisu's comment is referring to their alternative implementation, not the one presented in this PR.

**Comment by Deokishisu** [comment]
image
In the [DMA Discord](https://discord.com/channels/237788642238660610/237788642238660610/1273701149211361372), I asked Sophia to test the Union Room under the aforementioned conditions:

>  If the player saves in the 2nd floor of the Pokemon Center and resets, they can still enter the Union Room without going downstairs to run `Task_InitUnionRoom`.

`Task_InitUnionRoom` does **not** need to be run in order to enter the Union Room. I therefore suggest that my implementation [described further up](https://github.com/rh-hideout/pokeemerald-expansion/pull/5050#issuecomment-2254579421) be used instead of my worse, earlier implementation that this PR is currently built off of.

**Comment by AlexOn1ine** [comment]
> This is certainly more elegant than my current PR, but it costs a flag. I imagine a flag less version with the same end result would have an easier time getting merged from a Senate perspective

I don't think this would be blocking. Just make a define.

**Comment by Deokishisu** [comment]
> > This is certainly more elegant than my current PR, but it costs a flag. I imagine a flag less version with the same end result would have an easier time getting merged from a Senate perspective
> 
> I don't think this would be blocking. Just make a define.

[Like I said further up](https://github.com/rh-hideout/pokeemerald-expansion/pull/5050#issuecomment-2254587496), this can be easily and quickly refactored to use a regular temp flag instead of a brand new one. It is barely a change to do so. I didn't do that in my CD implementation because it was unnecessary.

**Comment by pkmnsnfrn** [comment]
@AlexOn1ine / @Bassoonian do I need to change this PR to be the "better" Crystal Dust implementation? I don't want to waste time if you prefer the old one.

**Comment by AlexOn1ine** [comment]
> @AlexOn1ine / @Bassoonian do I need to change this PR to be the "better" Crystal Dust implementation? I don't want to waste time if you prefer the old one.

Ye, I do think it would be better.

**Comment by pkmnsnfrn** [comment]
superceded by https://github.com/rh-hideout/pokeemerald-expansion/pull/5448
